app/service/payment_generator_service.rb

```diff
@@ -0,0 +1,66 @@
+class PaymentGeneratorService
+  def initialize(loan:)
+    @loan = loan
+    @amount = @loan.amount
+    @terms = @loan.terms
+    @rate = @loan.interest_rate
+    @date = @loan.day
+    @interest_only = @loan.interest_only
+  end
+
+  def create
+    years = terms / 12
+    effected_term = interest_only ? terms - 3 : terms
+    payment_monthly = total_monthly_payment(amount, rate, years, effected_term)
+    (0...terms).each do |term|
+      payment =
+        if interest_only and term < 3
+          interest = interest_monthly(amount, rate, terms).round(2)
+          loan.payments.build(
+            start_balance: amount,
+            interest: interest,
+            total_pmt: interest,
+            principal: 0.00,
+            day: date.next_month,
+            end_balance: amount
+          )
+        else
+          interest = interest_monthly(amount, rate, effected_term).round(2)
+          principal = (payment_monthly - interest).round(2)
+          end_balance = (amount - principal).round(2) >= 0 ? (amount - principal).round(2) : 0
+          payment = loan.payments.build(
+            start_balance: amount,
+            interest: interest,
+            total_pmt: payment_monthly.round(2),
+            principal: principal,
+            day: date.next_month,
+            end_balance: end_balance
+          )
+          self.amount = end_balance
+          payment
+        end
+      self.date = date.next_month
+      payment.save
+    end
+  end
+
+  def interest_monthly(current_start_balance, annual_interest_rate, terms)
+    (current_start_balance * (annual_interest_rate / (terms * 100)))
+  end
+
+  private
+
+  attr_reader :terms, :rate, :interest_only, :loan
```

Comment:
I don't think this is the position for the accessors, there are some options here:
1. Use the accessors if you want to access the variables outside the service, like:
```
payment_generator = PaymentGeneratorService.new(loan: @loan)
payment_generator. amount
```
If you don't want that then just use inside the service the variables with the `@` (and remove the accessors)

Also, the normal place for this accessor is above the `def initialize` method, not below private 🤔